TNL-5023: Add seeding and testing of large thread.

robrap opened the pull request:
### Description
 
[TNL-5023](https://openedx.atlassian.net/browse/TNL-5023)

TNL-5023: Forums Performance: Investigate Large Threads

Here is a first pass at seeding/testing large threads.  

This code adds a Locust Task Set for:

1. Seeding a large thread.
2. Reading the large thread.

The seeding was implemented as a Locust Task Set which can be run as a separate test.  This enables the seeding to take advantage of the parallelism of Locust, and we can also compare seeding times.

Here's how to seed:
```
BASIC_AUTH_USER=XXX BASIC_AUTH_PASSWORD=YYY COURSE_ID=course-v1:edX+DemoX+Demo_Course LOCUST_TASK_SET=SeedForumsTasks LOCUST_MIN_WAIT=1000 LOCUST_MAX_WAIT=1000 locust -c 10 -r 1 -n 4500 -H https://discussions.sandbox.edx.org --no-web
````

Testing against a large thread just does repeated reads:
```
BASIC_AUTH_USER=XXX BASIC_AUTH_PASSWORD=YYY LARGE_TOPIC_ID=i4x-edx-eiorguegnru-course-foobarbaz LARGE_THREAD_ID=579f8b6948ee34412e000078 COURSE_ID=course-v1:edX+DemoX+Demo_Course LOCUST_TASK_SET=ForumsTasks LOCUST_MIN_WAIT=1000 LOCUST_MAX_WAIT=1000 locust -c 10 -r 1 -H https://discussions.sandbox.edx.org --no-web
```

I'll drops some comments in the code as well.

### Reviewers (any 2 of the following):
- [x] @tobz 
- [x] @pwnage101 
- [ ] @ormsbee 
- [ ] @dianakhuang 

FYI: @andy-armstrong @bjacobel 

### Post-review
- [x] Squash commits

robrap commented on lms/forums.py:
```diff
@@ -1,16 +1,24 @@
 _thread_ids = deque(maxlen=1000)
 
+# track the large thread tuple (topic_id, thread_id)
+_large_thread = []
```
Reviewers: Is it fine to just continue this pattern used for _thread_ids?  Should these be moved to be class variables?

pwnage101 replied:
Fine by me.  But if you do move it into the class, remember to declare it global.

Also, why is `_large_thread` of type list if you never append more than one element or reference anything except the zeroth index?  AFAICT, you can initialize it as `None` and refer to it as a tuple rather than a list.

robrap replied:
@pwnage101: I feel like that caused an issue, but not sure when, so let me fix.

robrap replied:
I needed to add a `global` to make it work that I must have missed my first attempt.

robrap commented on lms/forums.py:
```diff
@@ -172,3 +180,139 @@ def followed_threads(self):
+
+class SeedForumsTasks(LmsTasks):
+
+    def create_thread(self, topic_id):
```
Reviewers: Some of this code is partly redundant with ForumsTasks.  Python question: if I want to share, since these are using `self.post()`, would I just add in a ForumsBaseTasks to inherit from?

pwnage101 replied:
I'm okay with leaving this as is, since the same-named tasks aren't exactly the same code.   If I were to ask for anything more, I'd probably ask that you somehow improve LmsTasks to reduce boilerplate code across all the tasks.

robrap replied:
@pwnage101: I did end up refactoring this before seeing your comment.  I moved shared code to a Base class.

I'm not sure what boilerplate code you are looking for in LmsTasks.  Code that affects this Forums code?  Otherwise, I'd rather just concentrate on the Forums code for this PR.

robrap commented on lms/forums.py:
```diff
@@ -172,3 +180,139 @@ def followed_threads(self):
+        )
+
+        # clump comments on fewer or responses
+        if not _large_thread_response_ids or random.randint(1, 10) > 9:
```
@tobz: How about this?
```
if not _large_thread_response_ids or random.randint(1, len(_large_thread_response_ids)) <= 1
    ...
```

tobz replied:
What happens if there are no threads? Does `not _empty_list_here` evaluate to a truthy expression?

robrap replied:
@tobz: Yes.  If it is empty, then the first response will get added and we don't make it to the second half of the or.

tobz commented on lms/forums.py:
```diff
@@ -172,3 +180,139 @@ def followed_threads(self):
+        )
+
+        # clump comments on a few responses
+        if not _large_thread_response_ids or random.randint(1, len(_large_thread_response_ids)) <= 1:
```
Can we just go ahead and change `not _large_thread_response_ids` to something like `len(_large_thread_response_ids) < RESPONSES_ABOVE_THE_FOLD` or something?  I don't really care what the constant name is, but I was mostly thinking in the sense of... how many responses are visible for most users without having to scroll down at all?

I think we're golden here if we have both  the "always add the first N responses" and "gradually comment less and less on newer responses" behavior.

robrap replied:
@tobz: I "think" I sort of understand what you are asking for, but not certain.  Do you want a bunch of comments guaranteed on the oldest response?  If so, I think this would require tracking the number of comments added, and then adding to the response ids after we have that number of comments first.  It feels a bit complicated, but may be needed.  Can we start with what I have and then see where it gets us?

pwnage101 commented on lms/forums.py:
```diff
@@ -172,3 +180,139 @@ def followed_threads(self):
+            LOGGER.info('To run ForumsLargeThreadTasks, use:\nLARGE_TOPIC_ID={} and LARGE_THREAD_ID={}'.format(
+                    new_thread[0], new_thread[1])
+            )
+            LOGGER.info('To run ForumsTasks, update course_data discussion_ids to:\n{}'.format(all_topic_ids))
```
How long does seeding take for a large thread?  Is this seeding process now required for ForumsTasks (not for just ForumsLargeThreadTasks), as implied by the output message?

robrap replied:
I know.  This wasn't clear because I was dumping course_data discussion ids, because this wasn't done anywhere else and I wanted to give someone the ability to fix the test (which is otherwise broken for LMS in general).  You don't actually require the large thread for lms tests, and it wasn't originally integrated.

I did add some integration where if you provide the large topic/thread, then LMS tests will also do some of its reads against that large thread.  Right now the ration is 45:10 (any thread:large thread).  We could adjust.  I just picked anything to get integration started.

pwnage101 replied:
high level question: why even have a ForumsLargeThreadTasks task set at all if the ForumsTasks can cover large threads now?

robrap commented:
@tobz @pwnage101: I made a pass at fixes.  Let me know if you have any required changes or how to get thumbs.  Thanks!

pwnage101 commented on lms/forums.py:
```diff
@@ -172,3 +267,83 @@ def followed_threads(self):
+        """
+        Read the large thread.
+        """
+        super(ForumsTasks, self).large_thread()
```
s/ForumsTasks/ForumsLargeThreadTasks/

behavior is the same, but it's more understandable.

robrap replied:
@pwnage101: Just let me know when you are done with your pass and I'll push changes.  Thanks.

pwnage101 commented on lms/forums.py:
```diff
@@ -23,7 +31,103 @@ def _dummy_text(minlen, maxlen):
+        Read the large thread.
+        """
+        discussion_id = self.large_topic_id
+        thread_id = self.large_thread_id
```
lines 85,86---no reason to do this really

pwnage101 commented on lms/forums.py:
```diff
@@ -133,8 +223,20 @@ def single_thread(self):
+
+        else:
+            # Fall back to regular read if not configured for long thread test
+            self.single_thread
```
does the fallback mode actually call self.single_thread, or does it just return a function object?

robrap replied:
Interesting.  This should be a function call.  However, I thought I tested this and saw single_threads appear in locust output, so I'm not sure what happened.

Either way, I will update and retest.  Thanks for catching this.

pwnage101 commented:
high level question: why even have a ForumsLargeThreadTasks task set at all if the ForumsTasks can cover large threads now?

pwnage101 commented:
@robrap 
> I'm not sure how to make this configurable.

In the usual way: shell variables.

Sorry I'm being critical, but why would you want the ability to crank up the ratio of one task (large threads)?  Do the results of that tell us anything new?  Why not just run the ForumsTasks with the ratios in this PR all the time?  Running the seed tasks would "configure" whether to test large threads or not.

Just making sure we don't add code where it isn't needed, but please correct me if I'm wrong.

robrap commented:
@pwnage101: It's a reasonable question and what you say makes sense.  I think this is just an artifact of how I built this, where integration to the other tests at all was a last minute idea.

I'm going to adjust the ratio to 30 single : 25 large and remove the Large Thread Task Set.  I have a future task to look at load numbers during our peak, and may adjust again at that point.

robrap commented:
@pwnage101: On a side note, Locust complains when you call interrupt() on the top-level Task Set, so I'm not sure of the best way to handle that for Task Sets that are sometimes run directly.  I guess instead we'd need LMS Task Sets with single sub-Task Sets, so it remains a sub-Task Set.

robrap commented:
@pwnage101 @tobz: I added final fixes and adjusted weightings based on my research.  This is ready for review.  I haven't actually run the latest latest changes, but I plan to shortly with @tobz for our current round of load testing against the sandbox.

robrap commented:
@pwnage101: Do you think we should move away from the hard coded course_data, and use the same call during setup of ForumsTasks to determine discussion ids?  I may do this is a separate PR.  Thoughts?

pwnage101 commented on lms/forums.py:
```diff
@@ -1,16 +1,24 @@
+# track the large thread tuple (topic_id, thread_id)
+_large_thread = None
+# track responses on the large thread
+_large_thread_response_ids = deque(maxlen=100)
```
I think there's no reason to declare `_large_thread` to be global in the SeedForumsTasks class below because it's already global to this module.   I think what I was talking about earlier was to move declaration and initialization of both of these variables into the various classes below as global member variables (e.g. `self._large_thread`).

I suppose you'd declare `self._large_thread` in BaseForumsTasks and `self._large_thread_response_ids` in SeedForumsTasks.

pwnage101 replied:
Okay, so I did a bit more research and it turns out class member variables can't be declared global anyway.  They're already sort of global when referenced like a module, e.g. `BaseForumsTask._large_thread`.  I'll leave it up to you how to move forward, but at least remove the global declaration and make sure that is a noop.

robrap commented:
@pwnage101 @tobz: I did a big refactor in the last commit.
1. Moved to class variables.
2. Removed discussion_ids (topic_ids) from course_data and dynamically load at the start of a test.  This make it much simpler to run the ForumsTasks against different courses in different environments.

@pwnage101: FYI: Moving to class variables, I no longer need the `global` call.  However, it was originally needed to avoid the following error:
```
UnboundLocalError: local variable '_large_thread' referenced before assignment
```

robrap commented:
@tobz @pwnage101: Sorry for the late commit (see above).  I'd like to see if I can get this merged before end of iteration (tomorrow morning), but that may not be possible.

FYI: @dianakhuang

pwnage101 commented on lms/forums.py:
```diff
@@ -127,40 +155,31 @@ def create_comment(self, response_id):
+
+    # keep track of thread ids created during a test, so they can be used to formulate read requests and replies.
+    # a deque is used instead of a list in order to enforce maximum length.
+    _thread_ids = deque(maxlen=1000)
```
can you add another comment about the intention of using this as a global variable, rather than an instance var?

robrap replied:
Done.

tobz commented on lms/forums.py:
```diff
@@ -23,50 +22,174 @@ def _dummy_text(minlen, maxlen):
+                for child in topic['children']:
+                    BaseForumsTasks._all_topic_ids.append(child['id'])
+
+        return BaseForumsTasks._all_topic_ids
```
This is great.  I think this will solve the problem I was having where the hardcoded `course` topic ID was making things funky.  It's also a nice sanity check that we make the discussion service tell us what it thinks it has.

tobz commented on lms/forums.py:
```diff
@@ -97,32 +220,17 @@ def create_thread(self):
         Read a specific thread.
         """
-        if not _thread_ids:
+        if not ForumsTasks._thread_ids:
```
Why can't this be `self._thread_ids`? (I'm not asking in the sense of "why didn't you do that?" but "is that not a valid way?")

robrap replied:
I tried and it didn't work. I'm new to Python, so not sure whether this should be a valid way to get to a class variable and I'm just doing something wrong?

pwnage101 replied:
Quoting Robert Raposa (2016-08-10 18:34:07)
> I tried and it didn't work. I'm new to Python, so not sure whether this
> should be a valid way to get to a class variable and I'm just doing something
> wrong?

IIRC, you're using `ForumsTasks._thread_ids` because it is shared between
instances of the class, and that's the one that gets defined in on_start.  If
you use the instance variable `self._thread_ids`, then it should throw
"undefined".

Troy

tobz commented:
👍

pwnage101 commented:
:+1:
